### PR TITLE
Replacing Deprecated Finalize() with autoCloseable from JAVA_API

### PR DESCRIPTION
**Pull request by WWW0030**
# Description

Replacing Deprecated `Finalize()` with `autoCloseable` from JAVA_API

Fixes # (issue) 
https://github.com/kuzudb/kuzu/issues/2974#event-14662134856

# Contributor agreement

- [ ] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

**Review by mewim — CHANGES_REQUESTED** [review]
There is a major issue in this PR. The `AutoCloseable` will only call `close()` automatically "when exiting a try-with-resources block for which the object has been declared" (as mentioned in Java docs). There is NO implication with garbage collection activities. But you have decided to remove all the manual calls to destroy the objects, which causes a lot of memory leak. I agree that we should not have both `destroy` and `close` public methods (which can be confusing). But instead of removing all the calls to `destroy`, you should change it to `close`.

File: tools/java_api/src/main/java/com/kuzudb/Connection.java [hunk 1]
```diff
@@ -31,19 +31,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/Connection.java:34]: The description here should be more descriptive.

File: tools/java_api/src/main/java/com/kuzudb/DataType.java [hunk 2]
```diff
@@ -30,19 +30,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/DataType.java:33]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/Database.java [hunk 3]
```diff
@@ -60,19 +60,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/Database.java:63]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/FlatTuple.java [hunk 4]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+     * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/FlatTuple.java:20]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java [hunk 5]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java:20]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/QueryResult.java [hunk 6]
```diff
@@ -18,11 +18,11 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+     * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/QueryResult.java:21]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/Value.java [hunk 7]
```diff
@@ -52,11 +52,11 @@ public void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+     * Close.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/Value.java:55]: Dittto

**Review by mewim — CHANGES_REQUESTED** [review]
Also, see my comments. All the `close()` function needs to have a descriptive comment. Since they are public methods, the comment will be used in generating the javadoc.

File: tools/java_api/src/main/java/com/kuzudb/Connection.java (hunk 1 above)
Comment (WWW0030) [reply tools/java_api/src/main/java/com/kuzudb/Connection.java]: Changed, please let me know if the new description is more fitting 👍

**Review by mewim — COMMENTED** [review]
I think the comments should be more specific. For example:
`Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.`

File: tools/java_api/src/test/java/com/kuzudb/ConnectionTest.java [hunk 8]
```diff
@@ -18,9 +18,7 @@ public class ConnectionTest extends TestBase {
 
     @Test
     void ConnCreationAndDestroy() {
-        try {
-            Connection conn = new Connection(db);
-            conn.destroy();
+        try ( Connection conn = new Connection(db);) {
```
Comment (mewim) [inline tools/java_api/src/test/java/com/kuzudb/ConnectionTest.java:21]: Fix the format here. It should not requre `;`.

File: tools/java_api/src/test/java/com/kuzudb/DatabaseTest.java [hunk 9]
```diff
@@ -15,35 +15,35 @@ public class DatabaseTest extends TestBase {
     void DBCreationAndDestroyWithArgs() {
         try {
             String dbPath = tempDir.toFile().getAbsolutePath();
-            Database database = new Database(
+            try (Database database = new Database(
```
Comment (mewim) [inline tools/java_api/src/test/java/com/kuzudb/DatabaseTest.java:18]: I think you should add seperate `catch` for database creation. If your database creation failed, your test case should be aware that it is caused by database creation failure. 
So something like:
```java
String dbPath;
try {
  String dbPath = tempDir.toFile().getAbsolutePath();
} catch(Exception e) {
  fail("Cannot get database path: " + e.getMessage();
}

try (Database database = new Database(...){
}
catch(e){
  fail("DBCreationAndDestroyWithArgs failed: " + e.getMessage();
}
```

**Review by mewim — CHANGES_REQUESTED** [review]
See my comments.

File: tools/java_api/src/test/java/com/kuzudb/DatabaseTest.java (hunk 9 above)
Comment (mewim) [reply tools/java_api/src/test/java/com/kuzudb/DatabaseTest.java:18]: Note that this applies to all functions in the same file

File: tools/java_api/src/main/java/com/kuzudb/DataType.java [hunk 10]
```diff
@@ -30,19 +30,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/DataType.java:33]: My example is for `Connection` class. You should change it for each class and not blindly copy it everywhere.

File: tools/java_api/src/main/java/com/kuzudb/Database.java [hunk 11]
```diff
@@ -60,19 +60,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/Database.java:63]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/FlatTuple.java [hunk 12]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/FlatTuple.java:20]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java [hunk 13]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java:20]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/QueryResult.java [hunk 14]
```diff
@@ -18,11 +18,11 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/QueryResult.java:21]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/Value.java [hunk 15]
```diff
@@ -52,11 +52,11 @@ public void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+    * Close the connection and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/Value.java:55]: Ditto

**Review by mewim — CHANGES_REQUESTED** [review]
See my comments

File: tools/java_api/src/main/java/com/kuzudb/DataType.java (hunk 10 above)
Comment (WWW0030) [reply tools/java_api/src/main/java/com/kuzudb/DataType.java:33]: Yes, sorry didn't check the comment closely enough, will fix

File: tools/java_api/src/main/java/com/kuzudb/QueryResult.java [hunk 16]
```diff
@@ -18,11 +18,11 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+    * Close the queryResult and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/QueryResult.java:21]: Using `camelCase` does not make sense here. It should either be `query result` (plain English) or `QueryResult` (consistent with the class name)

File: tools/java_api/src/main/java/com/kuzudb/DataType.java [hunk 17]
```diff
@@ -30,19 +30,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close the dataType and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/DataType.java:33]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/FlatTuple.java [hunk 18]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-     * Finalize.
+    * Close the flatTuple and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/FlatTuple.java:20]: Ditto

File: tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java [hunk 19]
```diff
@@ -17,19 +17,19 @@ private void checkNotDestroyed() throws ObjectRefDestroyedException {
     }
 
     /**
-    * Finalize.
+    * Close the perparedStatement and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
```
Comment (mewim) [inline tools/java_api/src/main/java/com/kuzudb/PreparedStatement.java:20]: Ditto

**Review by mewim — CHANGES_REQUESTED** [review]
See my comments

File: tools/java_api/src/main/java/com/kuzudb/QueryResult.java (hunk 16 above)
Comment (mewim) [reply tools/java_api/src/main/java/com/kuzudb/QueryResult.java:21]: If you decide to change it to `QueryResult`, you need to change `database` -> `Database` etc. to be consistent in other comments

Comment (mewim) [reply tools/java_api/src/main/java/com/kuzudb/QueryResult.java:21]: So I'd suggest to just go with `query result`

**Review by mewim — APPROVED** [review]
Can be merged if CI passes

**Pull request merged**